### PR TITLE
Added ability to select and pop tags. Added functionality to auto-merge Head tags if multiple present

### LINKED ISSUE #38: Double <head> tags
Currently, if you provide `html_head` to the `RapidHTML` class and then one of your route returns a `Head()` tag, you'll get two `<head>` tags in the `<html>` tag which is probably bad practice

This can be seen here https://github.com/ONTechNITS/RapidHTML/tree/master/examples/modern_table

We need to look for a `<head>` tag in a constructed `Html()` object and merge the contents of `Head()` and `html_head` together

### PR DESCRIPTION
**Pull request by rashidnhm**
This PR allows us to merge the `<head>` tag in responses if multiple of them are present.

It further adds a new feature where you can "crawl" and fetch elements within a page.

Sample usage:
```python
    test_page = Html(
        Head(
            Style(
                StyleSheet(h1={"color": "red"}, button={"color": "blue"})
            )
        ),
        Div(
            H1("BYE, world!"),
            Button("Click me", id="button", hx_get="/data"),
        ),
        A(
            "This is a hyperlinke to asdf.com",
            href="https://google.com",
            target="_blank",
        ),
        *[Hr(style="color: red") for _ in range(10)],
        Br(),
        P("Some more text"),
    )
```

Now if you wanted to get all the line elements (`<hr />`), you could do so by doing:
```python
hrs = test_page.select(Hr)
print(hrs)
# [<rapidhtml.tags.Hr object at 0x70f384d76920>, ...]
```

You can recursively find stuff too!
```python
h1_headers = test_page.select(H1, recurse=True)
```

There is also a `pop` and `first_match` argument.

Everything is documented within the code, tests were added as well.

Closes #38

### CODE REVIEW

File: src/rapidhtml/tags.py [hunk 1]
```diff
@@ -126,6 +142,99 @@ def render(self):
...
+            raise TypeError(f"Unexpected tag type {type(select_tag)}. {select_tag}")
+
+        ret_tags: list["BaseTag"] = []
+        self_matches: list[tuple[int, "BaseTag"]] = []
+
+        for i, tag in enumerate(self.tags):
+            if not isinstance(tag, BaseTag):
+                continue
+            tag_name = tag.tag_name
+            if tag_name == select_tag_name or tag_name == f"htmltag{select_tag_name}":
```
Comment (ThomasJRyan) [inline src/rapidhtml/tags.py:187]: As part of my callback PR, I actually normalized these tag names, so the latter part of this check isn't necessary
https://github.com/ONTechNITS/RapidHTML/blob/ec8ece5f9e1fb8032d7a4a90f17d024c646cc1bf/src/rapidhtml/tags.py#L50

File: src/rapidhtml/tags.py [hunk 2]
```diff
@@ -126,6 +142,99 @@ def render(self):
...
+                self_matches.append((i, tag))
+                if first_match:
+                    break
+            elif recurse and isinstance(tag, BaseTag):
+                sub_ret_tags = tag.select(
```
Comment (ThomasJRyan) [inline src/rapidhtml/tags.py:192]: It's an edge case, but if we had tags of the same type nested inside each other (I can only really think of `div` that would do this), I don't think this logic would get the parent?

File: src/rapidhtml/tags.py [hunk 3]
```diff
@@ -126,6 +142,99 @@ def render(self):
...
+            BaseTag or T: The selected tag or the default value if no matching tag is found.
+
+        Raises:
+            KeyError: If no matching tags are found and no default value is provided.
+            ValueError: If an instantiated object is passed as the tag parameter.
+            TypeError: If the tag parameter is of an unexpected type.
+
+        """
+        if default and len(default) > 1:
+            raise TypeError("pop expected at most 2 arguments, got more")
```
Comment (ThomasJRyan) [inline src/rapidhtml/tags.py:230]: I'm missing the reason why we're even allowing more than one value...

**Review by ThomasJRyan — CHANGES_REQUESTED** [review]

File: src/rapidhtml/tags.py (hunk 3 above)
Comment (ThomasJRyan) [reply src/rapidhtml/tags.py:230]: Ahhhhhhhh! That's weird but I like it! I never even thought about how that actually functioned. Well thought out

Comment (rashidnhm) [reply src/rapidhtml/tags.py:230]: But when using `*args`, but cannot restrict the user to how many params they pass, but check the values on the other end .. which is exactly what dict.pop does as well!

```bash
$ python -c '{"a": 1}.pop("b", None, None, None, None)'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: pop expected at most 2 arguments, got 5

```

File: src/rapidhtml/tags.py (hunk 2 above)
Comment (rashidnhm) [reply src/rapidhtml/tags.py:192]: Sir, I may need you to draw out this edge case a bit more. I am not following you on this

Comment (ThomasJRyan) [reply src/rapidhtml/tags.py:192]: So, I think my initial review was a bit off (I hadn't actually tested it yet), but looking at it now I think it's still a bit off

Say I have this:
```py
from rapidhtml.tags import *

html = Html(
    Div(
        Div(
            Div(
                id="3"
            ),
            id="2"
        ),
        id="1"
    )
)```
If I do `html.select(Div)` it seems I'll only get `[<rapidhtml.tags.Div at 0x7f048e55f730>]` which is the `Div` with `id="1"`

So tags nested inside other tags of the same type don't get selected. Though I can't think of any other tag that gets nested inside itself aside from `Div`, so it's a very low priority bug (who's trying to get all the `Div`s anyways?

Comment (rashidnhm) [reply src/rapidhtml/tags.py:192]: Ah I see what you are saying!

I thought about selecting tags based on the value of an attr, but I left it out for now, but I see this as an absolute enhancement on top of this that we should do!

Currently, if you knew the structure of the page, and wanted let's say Div 2, you'd have to do:

```python3
html.select(Div, first_match=True)[0].select(Div)
```

When you are selecting Div with Id1, it is returning the reference to the object, so you can subselect within it!

File: src/rapidhtml/tags.py [hunk 4]
```diff
@@ -126,6 +142,99 @@ def render(self):
...
+            raise TypeError(f"Unexpected tag type {type(select_tag)}. {select_tag}")
+
+        ret_tags: list["BaseTag"] = []
+        self_matches: list[tuple[int, "BaseTag"]] = []
+
+        for i, tag in enumerate(self.tags):
+            if not isinstance(tag, BaseTag):
+                continue
+            tag_name = tag.tag_name
+            if tag_name == select_tag_name or tag_name == f"htmltag{select_tag_name}":
```
Comment (rashidnhm) [inline src/rapidhtml/tags.py:187]:
```suggestion
            if tag.tag_name == select_tag_name:
```

File: src/rapidhtml/tags.py [hunk 5]
```diff
@@ -114,7 +130,7 @@ def render(self):
 
         # Recursively render child tags
         for tag in self.tags:
-            if isinstance(tag, (BaseTag, StyleSheet)):
+            if isinstance(tag, BaseTag):
```
[the remark this answers is not in the document]
Comment (rashidnhm) [reply src/rapidhtml/tags.py:150]: I don't mind the idea. It can be an abstract method so each class inheriting from it can define it's own thing for it

**Review by ThomasJRyan — APPROVED** [review]

**Pull request merged**